### DISCUSSION TITLE
아이템 18. 상속보다는 컴포지션을 사용하라

### DISCUSSION THREAD

**Discussion by corock** [Study-2-Effective-Java/Effective-Java#38] [3. 과제] [answered]
## TL;DR

- 상속은 캡슐화(*encapsulation*)를 깨뜨린다.
    - 상위 클래스(*superclass*)의 내부 구현이 달라질 때마다 하위 클래스(*subclass*)에 영향을 끼쳐 오동작을 유발할 수 있다.
- 합성(*composition*)은 특정 클래스 내 `private` 필드로 기존 클래스의 인스턴스를 참조하게 하는 것을 말한다.
    - `private` 필드로 참조하는 합성 관계의 클래스 인스턴스 메서드는 기존 클래스의 대응하는 메서드 호출을 통해 결과를 반환한다.
- 래퍼 클래스(*Wrapper Class*)는 하위 클래스보다 견고하고 강력하다.
- 상속은 상위 클래스와 하위 클래스가 순수한 *is-a* 관계일 때만 사용해야 한다.
    - 서브 타입(*subtype*)은 언제나 슈퍼 타입(*supertype*)으로 교체할 수 있어야 한다(a.k.a. LSP).

### 시나리오

- 우리에게 `HashSet` 을 사용하는 프로그램이 존재한다.
- 이 `HashSet` 은 처음 생성된 이후 원소가 몇 개 더해졌는지 알 수 있어야 한다.
- 따라서 변형된 `HashSet` 을 만들어 추가된 원소의 수를 저장하는 변수와 접근자 메서드를 추가했다.
- 그다음 `HashSet` 에 원소를 추가하는 메서드인 `add` 와 `addAll` 을 재정의했다. 

시나리오를 반영한 `InstrumentedHashSet` 클래스를 만들었다. 과연 잘 설계했다고 말할 수 있을까?

*InstrumentedHashSet.java*

```java
public class InstrumentedHashSet<E> extends HashSet<E> {

    // 추가된 원소의 수
    private int addCount = 0;

    public InstrumentedHashSet() {
    }

    public InstrumentedHashSet(int initCap, float loadFactor) {
        super(initCap, loadFactor);
    }

    @Override public boolean add(E e) {
        addCount++;
        return super.add(e);
    }

    @Override public boolean addAll(Collection<? extends E> c) {
        addCount += c.size();
        return super.addAll(c);
    }

    public int getAddCount() {
        return addCount;
    }

}
```

해당 클래스는 잘 구현된 것처럼 보이지만 제대로 작동하지 않는다.
이를 확인하기 위해 클라이언트 코드를 작성해보면 다음과 같다.

```java
InstrumentedHashSet<String> ihs = new InstrumentedHashSet<>();
ihs.addAll(List.of("틱", "탁탁", "펑"));
```

- 이제 `getAddCount` 메서드를 호출하면 `3` 을 반환하리라 기대하겠지만, 실제로는 `6` 을 반환한다.
- 원인은 `HashSet` 의 `addAll` 메서드가 `add` 메서드를 사용해 구현된 데 있다.
- 이 경우 하위 클래스에서 `addAll` 메서드를 재정의하지 않으면 문제를 고칠 수 있다.
- 하지만 당장은 제대로 동작할지 모르나, `HashSet` 의 `addAll` 이 `add` 메서드를 이용해 구현했음을 가정한 해법이라는 한계가 있다.

## 컴포지션을 활용하여 문제 해결하기

*InstrumentedSet.java*

- 다른 `Set` 인스턴스를 감싸고(*wrap*) 있다는 뜻에서 `InstrumentedSet` 같은 클래스를 래퍼 클래스라고 한다.
- 다른 `Set` 에 계측 기능을 덧씌운다는 뜻에서 데코레이터 패턴(*Decorator pattern*)이라고도 부른다.
- 래퍼 클래스는 콜백(*callback*) 프레임워크와 어울리지 않는다는 점을 제외하고 단점이 거의 없다.
  e.g. *Guava* 의 컬렉션 인터페이스용 전달 메서드

```java
public class InstrumentedSet<E> extends HashSet<E> {

    // 추가된 원소의 수
    private int addCount = 0;

    public InstrumentedSet(Set<E> s) {
        super(s);
    }

    @Override public boolean add(E e) {
        addCount++;
        return super.add(e);
    }

    @Override public boolean addAll(Collection<? extends E> c) {
        addCount += c.size();
        return super.addAll(c);
    }

    public int getAddCount() {
        return addCount;
    }

}
```

*ForwardingSet.java*

```java
public class ForwardingSet<E> implements Set<E> {

    private final Set<E> s;

    public ForwardingSet(Set<E> s)                               { this.s = s; }

    @Override public int size()                                  { return s.size(); }
    @Override public boolean isEmpty()                           { return s.isEmpty(); }
    @Override public boolean contains(Object o)                  { return s.contains(o); }
    @Override public Iterator<E> iterator()                      { return s.iterator(); }
    @Override public Object[] toArray()                          { return s.toArray(); }
    @Override public <T> T[] toArray(T[] a)                      { return s.toArray(a); }
    @Override public boolean add(E e)                            { return s.add(e); }
    @Override public boolean remove(Object o)                    { return s.remove(o); }
    @Override public boolean containsAll(Collection<?> c)        { return s.containsAll(c); }
    @Override public boolean addAll(Collection<? extends E> c)   { return s.addAll(c); }
    @Override public boolean retainAll(Collection<?> c)          { return s.retainAll(c); }
    @Override public boolean removeAll(Collection<?> c)          { return s.removeAll(c); }
    @Override public void clear()                                { s.clear(); }

}
```

**Comment by JoisFe** [comment]
TL;DR는 어떤 의미가 숨어있는 것 인가요?

**Reply by corock** [reply]
센스있는 질문이네요 😄

 *Too Long; Didn't Read.* 의 약어로 너무 길어서 읽고 싶지 않을 때 쓰는 속어로 알고 있습니다.
해당 아이템의 핵심을 요약해서 두괄식으로 배치하려고 했는데 "그래서 3줄 요약 좀" 같은 한국인의 정서를 반영해봤습니다.

**Reply by JoisFe** [reply]
아하 그렇군요
IUN(I understand now)

**Reply by YuDeokRin** [reply]
`TL;DR`  좋네요 나중에 활용해봐야겠군요

**Comment by JoisFe** [comment, marked as answer]
```상속보다는 구성(컴포지션)을 활용``` 디자인 원칙 중 정말 유명한 원칙이라 집중해서 정리하신 글을 읽었습니다.
특히 변형된 HashSet을 구현하는데 상속을 이용한 방법과 그로 인한 문제점을 구성으로 해결하는 예제가 이해하는데 매우 좋았던 것 같습니다.

```상속보다는 구성을 활용``` 한다는 것이 "A is a B" 보다 "A has a B"가 나을 수 있기 때문입니다. 한국말로 풀어보면 "A는 B이다" 보다는 "A에는 B가 있다"가 더 나을 수 있기 떄문입니다.

구성을 활용하는 방식이 디자인 패턴의 여러 패턴에서도 많이 쓰이기에 정말 중요하다고 알 고 있고 @corock 님께서 설명해주신 예시 뿐만 아니라 전략 패턴 (Strategy Pattern) 에서도 쓰여서 해당 예시를 참고하는 것도 좋은 것 같아 올립니다.

정말 유명한 예인데 추상 클래스 Duck과 추상메서드 quack()이 있을때 오리 종류마다 quack을 다르게 구현하고 싶은 상황이 있다고 가정합시다.

상속을 이용한다면 오리 종류마다 quack 메서드를 재정의 할 수 있으니 좋다고 판단하여 상속을 이용해 구현해보았습니다.
``` java
public abstract class Duck {
    abstract void quack();
}
```

``` java
public class JoisFeDuck extends Duck {

    @Override
    void quack() {
        System.out.println("꾸에에엑");
    }
}
```

``` java
public class CoRockDuck extends Duck {

    @Override
    void quack() {
        System.out.println("코엑");
    }
}
```

### 문제점
1. 오리가 소리내는 것 말고 나는 fly() 메서드 또한 추가 해야 하는 경우 상속 받는 모든 클래스에 fly() 메서드를 구현해야 함!! (현재는 JoisFeDuck, CoRockDuck 만 존재하지만 상속 받는 클래스가 많았다면 ??? 끔찍..) --> 유지 보수가 쉽지 않음..

2. 일부 상속받은 클래스는 fly() 메서드와 적합하지 않은 경우도 있을 것임

### 해결책
구성을 활용한 전략 패턴

### 전략 패턴 : 알고리즘군을 정의하고 캡슐화해서 각각의 알고리즘군을 수정해서 쓸 수 있게 해줌
- 전략 패턴을 사용시 클라이언트로 부터 알고리즘을 분리해서 독립적으로 변경 가능.

* 여기서 주의할 점은 여기서 캡슐화는 우리가 객체의 필드를 캡슐화하는 것과 다른 개념임
### 캡슐화 : 달리지는 부분(코드)을 찾아서 나머지 코드에 영향을 주지 않도록 분리하는 것.

캡슐화를 하면 바뀌지 않는 부분엥 영향을 미치지 않고 그 부분만 고치거나 확장할 수 있기에 유연한 코드가 됨!

그래서 전략 패턴을 적용해보면

``` java
public abstract class Duck {
    FlyBehavior flyBehavior;
    QuackBehavior quackBehavior;

    public Duck() {
    }
}
```

``` java
public class JoisFeDuck extends Duck {

    public JoisFeDuck() {
        super.quackBehavior = new JoQuack();
        super.flyBehavior = new FlyWithWings();
    }
}
```

``` java
public class CoRockDuck extends Duck {

    public CoRockDuck() {
        super.quackBehavior = new CoQuack();
        super.flyBehavior = new FlyNoWay();
    }
}
```

``` java
public interface QuackBehavior {

    void quack();
}
```

``` java
public class JoQuack implements QuackBehavior {

    @Override
    public void quack() {
        System.out.println("꾸에에엑");
    }
}

```

``` java
public class CoQuack implements QuackBehavior {

    @Override
    public void quack() {
        System.out.println("코엑");
    }
}
```

``` java
public interface FlyBehavior {

    void fly();
}
```

``` java
public class FlyWithWings implements FlyBehavior {

    @Override
    public void fly() {
        System.out.println("날 수 있어!!");
    }
}
```

``` java
public class FlyNoWay implements FlyBehavior {

    @Override
    public void fly() {
        System.out.println("못 날라 댕김");
    }
}
```

달라지는 부분인 quack, fly를 캡슐화 하여 각각의 알고리즘 군을 수정해서 사용할 수 있게됨
--> 유지 보수 가능

특히 Duck에서는 상속이 아닌 구성을 통해 QuackBehavior, FlyBehavior 구현체를 가지게 함으로써 시스템을 만들었기에 유연성을 크게 향상시킬 수 있게 되었습니다.

### 즉 구성을 이용하여 단순히 알고리즘군을 별도의 클래스 집합으로 캡슐화 하였고 구성 요소로 사용하는 객체에서 올바른 행동 인터페이스를 구현하기만 하면 실행 시에도 행동을 바꿀 수 있는 유연한 코드가 되었습니다.

* Reference : 페드퍼스트 디자인패턴 (한빛미디어)

**Reply by YuDeokRin** [reply]
답변이 거의 item 맡아서 쓴거 같네요 대단하십니다... ㄷㄷ

**Reply by corock** [reply]
@JoisFe 오우 디자인 패턴까지... 답변 잘 읽었습니다 :)
전략 패턴을 구현하는 과정에서 컴포지션을 활용하는 코드가 잘 이해되네요.

언급하신 대로 전략 패턴은 특정 행위(보통 *method* 겠죠?)를 클래스로 캡슐화합니다.
말 그대로 전략(*strategy*)을 쉽게 바꿀 수 있도록 해주는 *behavioral design pattern* 으로, 이렇게 하면 조건문을 줄일 수 있다는 장점이 있죠.

e-Commerce 도메인을 예로 들어보면, 주문 및 결제 시 회원 등급이나 프로모션 등 다양한 케이스에 대해 코드로 구현 시 `if-else` 절이 많이 생길 겁니다.
이때, 전략 패턴을 적극 도입함으로써 *conditional statement* 를 확 줄일 수 있는 효과가 있고 좋은 코드로 이어지는 방법의 하나로 알고 있어 공유합니다.
